Sming/Components/ssl/BearSsl/BrClientConnection.cpp

```diff
@@ -33,14 +31,93 @@ int BrClientConnection::init()
 	}
 
 	br_ssl_client_set_default_rsapub(&clientContext);
+	br_ssl_engine_set_x509(getEngine(), x509);
+	if(!br_ssl_client_reset(&clientContext, context.session.hostName.c_str(), 0)) {
+		debug_e("br_ssl_client_reset failed");
+		return getLastError();
+	}
 
-	// X509 verification
-	x509Context = new X509Context([this]() { return context.getSession().validateCertificate(); });
-	br_ssl_engine_set_x509(getEngine(), *x509Context);
+	return startHandshake();
+}
+
+template <class HashCtx> void resetHash(std::unique_ptr<HashCtx>& ctx, bool create)
```

Comment:
Could go in an anonymous namespace

Reply:
AFAIK templates won't be visible externally anyway, so am I missing something or is it just good practice?

Reply:
Thank you for pointing this out, noted for reference use of anonymous namespaces instead of static or internal stuff. https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sf22-use-an-unnamed-anonymous-namespace-for-all-internalnonexported-entities

Reply:
I'm not so sure about that. Take the following demo code, for example:
```cpp
// test.cpp
#include <iostream>
#include <typeinfo>

template <typename T>
void testFunc()
{
    std::cout << "testFunc called for type " << typeid(T).name() << "\n";
}

template void testFunc<int>();

```
```cpp
// main.cpp
#include <stdio.h>

template <typename T> void testFunc();

int main(int argc, char *argv[]) 
{
    testFunc<int>();
    
    return 0;
}
```
The command sequence below gives me no "undefined reference" error, so the explicitly instantiated template is visible to other translation. (Implicitly instantiated versions of `testFunc` are visible, too.)
```
g++ -o test.o -c test.cpp
g++ -o main.o -c main.cpp
g++ test.o main.o -o main
```
(BTW: clang++ gives the same result.)

Reply:
Thank you. I can confirm for this specific case as follows:

1. Amend template function declaration in `BrClientConnection.cpp` so it emits a symbol:

```
template <class HashCtx>
__attribute__((noinline)) void updateHash(HashCtx& ctx, const uint8_t* buf, size_t len)
```

2. Add some code to application.cpp:

```
namespace Ssl {
template <class HashCtx> void updateHash(HashCtx& ctx, const uint8_t* buf, size_t len);
}

void init()
{
  // ...
  std::unique_ptr<Crypto::Sha1> ctx;
  Ssl::updateHash(ctx, nullptr, 0);
}

```

And it compiles. Wrap the template declaration in an anonymous namespace and it doesn't.

Conclusion: anonymous namespace a good idea.